### PR TITLE
Clarifications on some of the specification

### PR DESCRIPTION
**Pull request by minrk**
- indicate that status=error doesn't require any fields from when status=ok to be defined
- mention that status messages can be used to fence output (with async output caveat)

closes jupyter/notebook#1576
closes jupyter/notebook#1575

### CODE REVIEW

**Comment by minrk** [comment]
cc @willingc for docs review.

File: docs/messaging.rst [hunk 1]
```diff
@@ -1052,9 +1064,26 @@ Message type: ``status``::
         execution_state : ('busy', 'idle', 'starting')
     }
 
+When a kernel receives a request and begins processing it,
+it shall immediately publish a status message with ``execution_state: 'busy'``.
+When that kernel has completed processing the request
+and has finished publishing associated IOPub messages, if any,
+it shall publish a status message with ``execution_state: 'idle'``.
+Thus, the outputs associated with a given execution shall generally arive
+between the idle and busy status messages associated with a given request.
```
Comment (takluyver) [inline docs/messaging.rst]: Flip round 'idle and busy' to 'busy and idle' for clarity, IMO.

File: docs/messaging.rst [hunk 2]
```diff
@@ -256,8 +257,37 @@ All messages sent to or received by any IPython process should have this
...
-===========================================
+Messages on the shell (ROUTER/DEALER) channel
+=============================================
+
+.. _request_reply:
+
+Request-Reply
+-------------
+
+In general, the ROUTER/DEALER sockets follow a request-reply pattern.
```
Comment (willingc) [inline docs/messaging.rst]: Suggest adding a colon at the end of this line and then bulleting the individual process steps.

File: docs/messaging.rst [hunk 3]
```diff
@@ -1052,9 +1064,26 @@ Message type: ``status``::
         execution_state : ('busy', 'idle', 'starting')
     }
 
+When a kernel receives a request and begins processing it,
+it shall immediately publish a status message with ``execution_state: 'busy'``.
```
Comment (willingc) [inline docs/messaging.rst]: it -> the kernel

File: docs/messaging.rst [hunk 4]
```diff
@@ -1052,9 +1064,26 @@ Message type: ``status``::
...
+When that kernel has completed processing the request
+and has finished publishing associated IOPub messages, if any,
+it shall publish a status message with ``execution_state: 'idle'``.
+Thus, the outputs associated with a given execution shall generally arive
+between the idle and busy status messages associated with a given request.
+
+.. note::
+
+    There is a caveat for asynchronous output,
```
Comment (willingc) [inline docs/messaging.rst]: **A caveat for asynchronous output**

Asynchronous output may be produced after the kernel has sent the idle status message. The kernel sends an idle message when the kernel has completed processing the request. The handling of asynchronous output messages is currently undefined in the specification (reference link to spec??). The currently implemented behavior is the Jupyter Notebook ...

**Comment by willingc** [comment]
Looks good to me. A few suggestions that may help the reader.

File: docs/messaging.rst (hunk 4 above)
Comment (minrk) [inline docs/messaging.rst]: This file is the spec, so I don't think there's a link to add, since it would be to here.

Comment (minrk) [inline docs/messaging.rst]: I've updated the text to "this specification" to hopefully clarify that.

**Comment by minrk** [comment]
Updated per review comments.

**Pull request merged**